### PR TITLE
Handle termination gracefully for controller manager and scheduler

### PR DESCRIPTION
**Pull request by mfojtik**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

This change will wire the stop channel baked by shutdown signal handler down to controller manager and scheduler. Doing this will cause these two properly close and release their ports used for serving connections.

This is causing problems if you run these in containers with host ports for example, where replacing old container with new container means you have to wait until kernel free up the TCP port for next process.

Credits to @sttts for most of this wiring :-)

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: cmd/kube-scheduler/app/server.go [hunk 1]
```diff
@@ -227,27 +238,12 @@ func Run(cc schedulerserverconfig.CompletedConfig, stopCh <-chan struct{}) error
...
-		case <-ctx.Done():
-		}
-	}()
-
 	// If leader election is enabled, runCommand via LeaderElector until done and exit.
 	if cc.LeaderElection != nil {
 		cc.LeaderElection.Callbacks = leaderelection.LeaderCallbacks{
-			OnStartedLeading: run,
+			OnStartedLeading: func(context.Context) {
+				sched.Run()
```
Comment (sttts) [inline cmd/kube-scheduler/app/server.go]: what does this mean when the context is closed? Will `leaderElector.Run(ctx)` below return ever if this call does not use the context?

File: cmd/kube-scheduler/app/server.go [hunk 2]
```diff
@@ -258,13 +254,12 @@ func Run(cc schedulerserverconfig.CompletedConfig, stopCh <-chan struct{}) error
...
-
-		return fmt.Errorf("lost lease")
+	} else {
+		// Leader election is disabled, so runCommand inline until done.
+		sched.Run()
 	}
 
-	// Leader election is disabled, so runCommand inline until done.
-	run(ctx)
-	return fmt.Errorf("finished without leader elect")
```
Comment (sttts) [inline cmd/kube-scheduler/app/server.go]: to be clear here: we change the return value to be inline with kcm and ccm.

File: staging/src/k8s.io/apiserver/pkg/server/secure_serving.go [hunk 3]
```diff
@@ -133,6 +133,8 @@ func RunServer(
 		return nil, fmt.Errorf("listener must not be nil")
 	}
 
+	doneCh := make(chan struct{})
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/secure_serving.go]: this was a rebase failure. We have stoppedCh. No need for doneCh.

File: cmd/kube-scheduler/app/server.go (hunk 1 above)
Comment (mfojtik) [reply cmd/kube-scheduler/app/server.go]:
```
sched.config.StopEverything
```

It seems it use this chan to synchronize over the provided context?

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 4]
```diff
@@ -37,6 +38,7 @@ import (
 	cliflag "k8s.io/component-base/cli/flag"
 	"k8s.io/component-base/cli/globalflag"
 	"k8s.io/klog"
+
```
Comment (hzxuzhonghu) [inline cmd/cloud-controller-manager/app/controllermanager.go]: remove

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 5]
```diff
@@ -109,6 +111,16 @@ the cloud specific control loops shipped with Kubernetes.`,
 
 // Run runs the ExternalCMServer.  This should never exit.
 func Run(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.TODO())
```
Comment (hzxuzhonghu) [inline cmd/cloud-controller-manager/app/controllermanager.go]: I think you want to use this ctx in belowing `leaderelection.RunOrDie`

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 6]
```diff
@@ -109,6 +111,16 @@ the cloud specific control loops shipped with Kubernetes.`,
...
+	defer cancel()
```
Comment (hzxuzhonghu) [inline cmd/cloud-controller-manager/app/controllermanager.go]: cancel is called too many times, pretty sure duplicate with line 166

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 7]
```diff
@@ -273,7 +273,7 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 // Run spawns the secure http server. It only returns if stopCh is closed
 // or the secure port cannot be listened on initially.
 func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
-	err := s.NonBlockingRun(stopCh)
+	serverDoneCh, err := s.NonBlockingRun(stopCh)
 	if err != nil {
```
Comment (stewart-yu) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: this change seem odd? `NonBlockingRun` return two `argument`?

**Review by stewart-yu — COMMENTED** [review]
BTW，fix some basical error in local

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 4 above)
Comment (mfojtik) [reply cmd/cloud-controller-manager/app/controllermanager.go]: done

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 5 above)
Comment (mfojtik) [reply cmd/cloud-controller-manager/app/controllermanager.go]: sure, done

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 6 above)
Comment (mfojtik) [reply cmd/cloud-controller-manager/app/controllermanager.go]: fixed

**Review by stewart-yu — COMMENTED** [review]
and where the change about ccm?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 8]
```diff
@@ -314,8 +317,12 @@ func (s preparedGenericAPIServer) NonBlockingRun(stopCh <-chan struct{}) error {
...
 		if err != nil {
 			close(internalStopCh)
-			return err
+			return nil, err
 		}
+	} else {
+		ch := make(chan struct{})
+		close(ch)
+		stoppedCh = ch
 	}
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: I think it does not, in `go` it's okay to read from the closed channel, see: https://golang.org/ref/spec#Close

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 9]
```diff
@@ -288,12 +288,15 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 	// Wait for all requests to finish, which are bounded by the RequestTimeout variable.
 	s.HandlerChainWaitGroup.Wait()
 
+	// wait for server listener to be closed
+	<-serverDoneCh
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: I think we could read from this chan before `s.HandlerChainWaitGroup.Wait()` so that we know that the lister has been closed.

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 10]
```diff
@@ -86,7 +86,7 @@ const (
 )
 
 // NewControllerManagerCommand creates a *cobra.Command object with default parameters
-func NewControllerManagerCommand() *cobra.Command {
+func NewControllerManagerCommand(stopCh <-chan struct{}) *cobra.Command {
```
Comment (p0lyn0mial) [inline cmd/kube-controller-manager/app/controllermanager.go]: commands are no longer accepting `stopCh` chan, instead, they register handlers for the signals when they need it - see: https://github.com/kubernetes/kubernetes/pull/76659

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 11]
```diff
@@ -115,7 +116,7 @@ controller, and serviceaccounts controller.`,
 				os.Exit(1)
 			}
 
-			if err := Run(c.Complete(), wait.NeverStop); err != nil {
```
Comment (p0lyn0mial) [inline cmd/kube-controller-manager/app/controllermanager.go]: do we want to wire signals for `cloud-controller-manager`?

https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/app/controllermanager.go#L78

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 12]
```diff
@@ -26,10 +26,12 @@ import (
 
 	"github.com/spf13/cobra"
 
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/uuid"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/apiserver/pkg/server"
+	genericapiserver "k8s.io/apiserver/pkg/server"
```
Comment (cheftako) [inline cmd/cloud-controller-manager/app/controllermanager.go]: This seems redundant with the prior line. Do we need an additional alias (genericapiserver) for a package we are already pulling in? We could just have server.SetupSignalHandler() below.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 13]
```diff
@@ -113,6 +115,15 @@ the cloud specific control loops shipped with Kubernetes.`,
 
 // Run runs the ExternalCMServer.  This should never exit.
 func Run(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.TODO())
+	go func() {
+		select {
+		case <-stopCh:
+			cancel()
+		case <-ctx.Done():
```
Comment (cheftako) [inline cmd/cloud-controller-manager/app/controllermanager.go]: What is the purpose of having both a stop channel and a done channel here? Especially as a context is usually associated with a request and our main run method does not seem related to a request.

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 14]
```diff
@@ -268,21 +283,23 @@ func Run(c *config.CompletedConfig, stopCh <-chan struct{}) error {
...
 		Lock:          rl,
 		LeaseDuration: c.ComponentConfig.Generic.LeaderElection.LeaseDuration.Duration,
 		RenewDeadline: c.ComponentConfig.Generic.LeaderElection.RenewDeadline.Duration,
 		RetryPeriod:   c.ComponentConfig.Generic.LeaderElection.RetryPeriod.Duration,
 		Callbacks: leaderelection.LeaderCallbacks{
 			OnStartedLeading: run,
 			OnStoppedLeading: func() {
-				klog.Fatalf("leaderelection lost")
+				cancel()
+				utilruntime.HandleError(fmt.Errorf("leaderelection lost"))
```
Comment (cheftako) [inline cmd/kube-controller-manager/app/controllermanager.go]: This needs to be fatal and this change appears to make it non fatal.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 15]
```diff
@@ -199,21 +214,23 @@ func Run(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan struct{}) error
...
 		Lock:          rl,
 		LeaseDuration: c.ComponentConfig.Generic.LeaderElection.LeaseDuration.Duration,
 		RenewDeadline: c.ComponentConfig.Generic.LeaderElection.RenewDeadline.Duration,
 		RetryPeriod:   c.ComponentConfig.Generic.LeaderElection.RetryPeriod.Duration,
 		Callbacks: leaderelection.LeaderCallbacks{
 			OnStartedLeading: run,
 			OnStoppedLeading: func() {
-				klog.Fatalf("leaderelection lost")
+				cancel()
+				utilruntime.HandleError(fmt.Errorf("leaderelection lost"))
```
Comment (cheftako) [inline cmd/cloud-controller-manager/app/controllermanager.go]: This needs to be fatal and this change appears to make it non fatal.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 16]
```diff
@@ -151,9 +163,13 @@ func Run(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan struct{}) error
 	if c.SecureServing != nil {
 		unsecuredMux := genericcontrollermanager.NewBaseHandler(&c.ComponentConfig.Generic.Debugging, checks...)
 		handler := genericcontrollermanager.BuildHandlerChain(unsecuredMux, &c.Authorization, &c.Authentication)
-		// TODO: handle stoppedCh returned by c.SecureServing.Serve
-		if _, err := c.SecureServing.Serve(handler, 0, stopCh); err != nil {
+		if serverStoppedCh, err := c.SecureServing.Serve(handler, 0, stopCh); err != nil {
```
Comment (p0lyn0mial) [inline cmd/cloud-controller-manager/app/controllermanager.go]: `Serving` should use `ctx.Done()`

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 12 above)
Comment (p0lyn0mial) [reply cmd/cloud-controller-manager/app/controllermanager.go]: yes, thx.

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 14 above)
Comment (p0lyn0mial) [reply cmd/kube-controller-manager/app/controllermanager.go]: no, this pull introduces a graceful shutdown for `kcm`, `ccm` and scheduler, in this case, it means that when the component loses leadership it notifies and waits for all dependant controllers and listeners before shutting down. For example, for `kcm` it means that it will wait for all its controllers as well as for HTTPS and HTTP listeners.

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 15 above)
Comment (p0lyn0mial) [reply cmd/cloud-controller-manager/app/controllermanager.go]: please see my previous comment https://github.com/kubernetes/kubernetes/pull/76452#discussion_r322244915

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 13 above)
Comment (p0lyn0mial) [reply cmd/cloud-controller-manager/app/controllermanager.go]: `stop` channel is controlled by the signals( SIGTERM and SIGINT) whereas `done` channel is controlled by `Run` method and allows for graceful termination. For example, `done` channel will be closed when the component cannot create `HTTP{S}` sockets, when it loses leadership or when it receives one of the signals. Note that closing one of the channels is equivalent to requesting closing the application.

**Comment by p0lyn0mial** [comment]
Alright, I think this pull is ready for review, PTAL.

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 14 above)
Comment (cheftako) [reply cmd/kube-controller-manager/app/controllermanager.go]: The fundamental guarantee we have for controllers right now is that we will not run them concurrently. (Or at least minimize then window where that might happen). This change violates that guarantee. 

As soon as we are told we are not the leader anymore (OnStopLeading). We have to assume that another KCM has taken over the leadership role. We also know that we have other threads in this process which are continuing the role of active controllers. They must be stopped immediately to prevent them from making concurrent changes with the new KCM master. This needs to be fatal.

Comment (cheftako) [reply cmd/kube-controller-manager/app/controllermanager.go]: FYI I'm not saying that having the process kill itself on OnStoppedLeading is the ideal solution for controller concurrency. I think we can do better. However I believe this kill itself behavior is needed to HA clusters until we build a better solution of controller concurrency.

File: cmd/kube-scheduler/app/server.go (hunk 1 above)
Comment (misterikkit) [reply cmd/kube-scheduler/app/server.go]: Storing a context or stop chan in a struct is generally not preferred. Passing the context into `Scheduler.Run` would be more idiomatic.

File: cmd/kube-scheduler/app/server.go [hunk 17]
```diff
@@ -213,55 +224,46 @@ func Run(cc schedulerserverconfig.CompletedConfig, stopCh <-chan struct{}, regis
...
-		}
-	}()
+	cc.InformerFactory.WaitForCacheSync(ctx.Done())
+	cache.WaitForCacheSync(ctx.Done(), cc.PodInformer.Informer().HasSynced)
 
 	// If leader election is enabled, runCommand via LeaderElector until done and exit.
 	if cc.LeaderElection != nil {
 		cc.LeaderElection.Callbacks = leaderelection.LeaderCallbacks{
-			OnStartedLeading: run,
+			OnStartedLeading: func(context.Context) {
```
Comment (misterikkit) [inline cmd/kube-scheduler/app/server.go]: So... Are we reusing the `sched` object each time this process. I don't see any code that would exit after graceful cleanup. I'm certain that attempting to re-use this object will fail. The leader election context is being ignored, and we still have a cancelled context in the sched struct. (This is why it would be preferable to pass context into `sched.Run()`.

**Review by misterikkit — COMMENTED** [review]
This does some wiring of context/stopCh into scheduler, but only at the surface level. There are some goroutines in the scheduler which do not propagate cancellation because we operate on the assumption that the process will exit when leadership is lost.

With this change, those goroutines could cause bad behavior by competing with the new leader to do writes. e.g.
1. old leader selects node A for pod
2. new leader selects node B for pod
3. new leader successfully binds pod
4. old leader fails to bind pod to node A, and updates pod status with SchedulingFailed.
5. mayhem

**Comment by misterikkit** [comment]
@ahg-g  FYI

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 18]
```diff
@@ -199,21 +213,23 @@ func Run(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan struct{}) error
...
-	leaderelection.RunOrDie(context.TODO(), leaderelection.LeaderElectionConfig{
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
 		Lock:          rl,
 		LeaseDuration: c.ComponentConfig.Generic.LeaderElection.LeaseDuration.Duration,
 		RenewDeadline: c.ComponentConfig.Generic.LeaderElection.RenewDeadline.Duration,
 		RetryPeriod:   c.ComponentConfig.Generic.LeaderElection.RetryPeriod.Duration,
 		Callbacks: leaderelection.LeaderCallbacks{
 			OnStartedLeading: run,
 			OnStoppedLeading: func() {
-				klog.Fatalf("leaderelection lost")
```
Comment (deads2k) [inline cmd/cloud-controller-manager/app/controllermanager.go]: I think that most of this pull is useful even without this change.  @mfojtik can you keep this fatal and solve the 90% case first?

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 14 above)
Comment (p0lyn0mial) [reply cmd/kube-controller-manager/app/controllermanager.go]: yes, we probably should, I didn't know that KCM has such strong assumptions in this area especially the scheduler. Although it seems like the leader election library doesn't [guarantee](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L17) anything - it may happen that two KCMs will be running at the same time. Can someone confirm this?

I suspect that KCM cares about efficiency - since correctness will be checked by the API server (`resourceVersion`). The scheduler, on the other hand, seems to care about correctness - https://github.com/kubernetes/kubernetes/pull/76452#pullrequestreview-285849073

Comment (cheftako) [reply cmd/kube-controller-manager/app/controllermanager.go]: Currently the KCMs in HA configurations use the fatal on leader election to ensure that there are not two active KCMs running at the same time. While it would be nice to get additional efficiency by letting multiple KCMs process simultaneously, I do not think we have the necessary correctness guarantees in place for that to be safe. resourceVersion is not sufficient for all controllers to behave correctly. (Eg. Fairly sure things like the cron/job controllers will schedule too much work)
